provider-service/base/pkg/server/resource/experiment_test.go

```diff
@@ -0,0 +1,158 @@
+)
+
+var _ = Describe("Experiment", Ordered, func() {
+	var (
+		mockProvider MockExperimentProvider
+		exp          Experiment
+	)
+
+	BeforeEach(func() {
+		mockProvider = MockExperimentProvider{}
+		exp = Experiment{Provider: &mockProvider}
+	})
+
+	Context("Create", func() {
+		When("valid json passed, and provider returns success", func() {
+			It("returns the id of the resource", func() {
+				ed := ExperimentDefinition{}
+				jsonExperiment, err := json.Marshal(ed)
+				Expect(err).ToNot(HaveOccurred())
+
+				id := "some-id"
+
+				mockProvider.On("CreateExperiment", ed).Return(id, nil)
+				response, err := exp.Create(jsonExperiment)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(response).To(Equal(ResponseBody{Id: id}))
+			})
+		})
+
+		When("invalid json is passed", func() {
+			It("errors", func() {
+				invalidJson := []byte(`/n`)
+
+				response, err := exp.Create(invalidJson)
+
+				Expect(err).To(HaveOccurred())
+				Expect(response).To(Equal(ResponseBody{}))
+			})
+		})
+
+		When("provider errors", func() {
+			It("errors", func() {
+				ed := ExperimentDefinition{}
+				jsonExperiment, err := json.Marshal(ed)
+				Expect(err).ToNot(HaveOccurred())
+
+				expectedErr := errors.New("some-error")
+
+				mockProvider.On("CreateExperiment", ed).Return("", expectedErr)
+
+				response, err := exp.Create(jsonExperiment)
+
+				Expect(err).To(Equal(expectedErr))
+				Expect(response).To(Equal(ResponseBody{Id: ""}))
+			})
+		})
+	})
+
+	Context("Update", func() {
+		When("valid json passed, and provider operations succeed", func() {
+			It("returns no error", func() {
+				ed := ExperimentDefinition{}
+				jsonExperiment, err := json.Marshal(ed)
+				Expect(err).ToNot(HaveOccurred())
+
+				id := "some-id"
+				updatedId := "some-update-id"
+				mockProvider.On("UpdateExperiment", ed, id).Return(updatedId, nil)
+				resp, err := exp.Update(id, jsonExperiment)
+
+				Expect(err).ToNot(HaveOccurred())
+				Expect(resp.Id).To(Equal(updatedId))
+			})
+		})
+
+		When("invalid json is passed", func() {
+			It("errors", func() {
+				invalidJson := []byte(`/n`)
+				resp, err := exp.Update("some-id", invalidJson)
+
+				Expect(err).To(HaveOccurred())
+				Expect(resp.Id).To(BeEmpty())
+			})
+		})
+
+		When("provider errors", func() {
+			It("errors", func() {
+				ed := ExperimentDefinition{}
+				jsonExperiment, err := json.Marshal(ed)
+				Expect(err).ToNot(HaveOccurred())
+
+				expectedErr := errors.New("some-error")
+				id := "some-id"
+				mockProvider.On("UpdateExperiment", ed, id).Return("", expectedErr)
+				resp, err := exp.Update(id, jsonExperiment)
+
+				Expect(err).To(Equal(expectedErr))
+				Expect(resp.Id).To(BeEmpty())
+			})
+		})
+	})
+
+	Context("Delete", func() {
+		When("valid id is passed and provider operations succeed", func() {
+			It("return no error", func() {
+				id := "some-id"
+				mockProvider.On("DeleteExperiment", id).Return(nil)
+				err := exp.Delete(id)
+				Expect(err).ToNot(HaveOccurred())
```

Comment:
```suggestion

				Expect(err).ToNot(HaveOccurred())
```

I actually prefer a space on both sides of the actual thing under test but dont think we have a standard

Reply:
https://github.com/sky-uk/kfp-operator/pull/435

before/after each assertion block rather than every assertion looks decent to me